usvm-ts/src/main/kotlin/org/usvm/machine/operator/TSBinaryOperator.kt

```diff
@@ -0,0 +1,511 @@
+            scope: TSStepScope,
+        ): UExpr<out USort> = with(lhs.tctx) {
+            check(lhs.isFakeObject() || rhs.isFakeObject())
+
+            scope.calcOnState {
+                val conjuncts = mutableListOf<ExprWithTypeConstraint<UBoolSort>>()
+                val groundFalseBranch = makeSymbolicPrimitive(boolSort)
+
+                if (lhs.isFakeObject() && rhs.isFakeObject()) {
+                    val lhsType = memory.typeStreamOf(lhs).single() as FakeType
+                    val rhsType = memory.typeStreamOf(rhs).single() as FakeType
+
+                    scope.assert(
+                        mkAnd(
+                            lhsType.mkExactlyOneTypeConstraint(ctx),
+                            rhsType.mkExactlyOneTypeConstraint(ctx)
+                        )
+                    )
+
+                    conjuncts += ExprWithTypeConstraint(
+                        constraint = mkAnd(lhsType.boolTypeExpr, rhsType.boolTypeExpr),
+                        expr = mkEq(
+                            memory.read(getIntermediateBoolLValue(lhs.address)),
+                            memory.read(getIntermediateBoolLValue(rhs.address))
+                        )
+                    )
+
+                    conjuncts += ExprWithTypeConstraint(
+                        constraint = mkAnd(lhsType.fpTypeExpr, rhsType.fpTypeExpr),
+                        expr = mkFpEqualExpr(
+                            memory.read(getIntermediateFpLValue(lhs.address)),
+                            memory.read(getIntermediateFpLValue(rhs.address))
+                        )
+                    )
+
+                    conjuncts += ExprWithTypeConstraint(
+                        constraint = mkAnd(lhsType.refTypeExpr, rhsType.refTypeExpr),
+                        expr = mkHeapRefEq(
+                            memory.read(getIntermediateRefLValue(lhs.address)),
+                            memory.read(getIntermediateRefLValue(rhs.address))
+                        )
+                    )
+
+                    conjuncts += ExprWithTypeConstraint(
+                        constraint = mkAnd(lhsType.boolTypeExpr, rhsType.fpTypeExpr),
+                        expr = mkFpEqualExpr(
+                            boolToFp(memory.read(getIntermediateBoolLValue(lhs.address))),
+                            memory.read(getIntermediateFpLValue(rhs.address))
+                        )
+                    )
+
+                    conjuncts += ExprWithTypeConstraint(
+                        constraint = mkAnd(lhsType.fpTypeExpr, rhsType.boolTypeExpr),
+                        expr = mkFpEqualExpr(
+                            memory.read(getIntermediateFpLValue(lhs.address)),
+                            boolToFp(memory.read(getIntermediateBoolLValue(rhs.address)))
+                        )
+                    )
+
+                    // TODO: support objects
+                }
+
+                if (lhs.isFakeObject()) {
```

Comment:
This condition always holds when the `lhs.isFakeObject() && rhs.isFakeObject()` holds. So, we need return or smth